File: src/routes/trading/control.py
# Copyright (c) 2022-2026 MKM Research Labs.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
Storm Sequence Control — GET/POST endpoints for storm_control.json.

Allows the Control tab in the trading desk to read and write the
centralised storm stress workflow parameters.
"""

from flask import jsonify, request

from config.auth import FUNC_TRADE_PRS, WRITE
from config.storm_control import (
    apply_storm_control,
    get_defaults,
    load_storm_control,
    save_storm_control,
)

from .._rbac import require
from .blueprint import trading_bp


@trading_bp.route("/trading/control/params", methods=["GET"])
def get_control_params():
    """Return current storm control parameters (from JSON or defaults)."""
    data = load_storm_control()
    if not data:
        data = get_defaults()
        source = "defaults"
    else:
        source = "json"
    return jsonify({"status": "success", "params": data, "source": source})


@trading_bp.route("/trading/control/params", methods=["POST"])
@require(FUNC_TRADE_PRS, WRITE)
def save_control_params():
    """Save updated storm control parameters and hot-reload.

    Admin-only: requires ``X-Admin-Password`` header matching the shared
    port/admin credential. Changes here affect every storm calculation
    across the platform, so gating prevents accidental writes.
    """
    body = request.get_json(silent=True)
    if not body:
        return jsonify({"status": "error", "message": "No JSON body"}), 400

    # Basic structural validation
    if "sections" not in body and "version" not in body:
        return jsonify({"status": "error", "message": "Missing sections"}), 400

    save_storm_control(body)
    apply_storm_control()
    return jsonify({"status": "success", "message": "Parameters saved and applied"})


@trading_bp.route("/trading/control/reset", methods=["POST"])
@require(FUNC_TRADE_PRS, WRITE)
def reset_control_params():
    """Reset storm control parameters to Python defaults.

    Admin-only: see ``save_control_params`` for rationale.
    """
    defaults = get_defaults()
    save_storm_control(defaults)
    apply_storm_control()
    return jsonify({"status": "success", "params": defaults})

File: src/static/js/storm/stormportfolio/control.js
// Copyright (c) 2022-2026 MKM Research Labs.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

            function addMapControl() {
                function findMap() {
                    var mapKey = Object.keys(window).find(function(k) { return k.startsWith('map_'); });
                    if (mapKey) return window[mapKey];
                    return null;
                }

                function tryAdd() {
                    var map = findMap();
                    if (!map) {
                        setTimeout(tryAdd, 500);
                        return;
                    }

                    var PortfolioControl = L.Control.extend({
                        options: { position: 'topright' },
                        onAdd: function() {
                            var container = L.DomUtil.create('div', 'leaflet-bar leaflet-control');
                            var btn = L.DomUtil.create('a', '', container);
                            btn.href = '#';
                            btn.title = 'Portfolio Storm Impact';
                            btn.setAttribute('role', 'button');
                            btn.innerHTML = '<svg width="20" height="20" viewBox="0 0 24 24" fill="none" stroke="#333" stroke-width="2"><path d="M2 12c1-3 3-5 5-5s3 2 5 2 3-2 5-2 4 2 5 5"/><path d="M2 17c1-3 3-5 5-5s3 2 5 2 3-2 5-2 4 2 5 5"/></svg>';
                            btn.style.cssText = 'display:flex;align-items:center;justify-content:center;width:30px;height:30px;cursor:pointer;background:white;';

                            L.DomEvent.disableClickPropagation(container);
                            L.DomEvent.on(btn, 'click', function(e) {
                                L.DomEvent.preventDefault(e);
                                showPanel();
                            });
                            return container;
                        }
                    });

                    new PortfolioControl().addTo(map);
                }

                setTimeout(tryAdd, 1000);
            }

            // Add map control button on load
            addMapControl();
